File: ROS2/src/learning_tf2_py/learning_tf2_py/static_turtle_tf2_broadcaster.py
# Copyright 2021 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys

# 这个是数据格式给到tf tree
from geometry_msgs.msg import TransformStamped

import rclpy
from rclpy.node import Node

from tf2_ros.static_transform_broadcaster import StaticTransformBroadcaster

# 提供欧拉角和四元数之间的转换
import tf_transformations


class StaticFramePublisher(Node):
    """
    Broadcast transforms that never change. 发布不会变化的左边关系所以是Static的

    This example publishes transforms from `world` to a static turtle frame.
    The transforms are only published once at startup, and are constant for all
    time.
    """

    def __init__(self, transformation):
        super().__init__('static_turtle_tf2_broadcaster')

        # 实例化一个 StaticTransformBroadcaster
        self._tf_publisher = StaticTransformBroadcaster(self)

        # Publish static transforms once at startup 这是一种静态的发布，在echo的时候只会出现一次，不会一直刷新的那种
        self.make_transforms(transformation)

    def make_transforms(self, transformation):
        static_transformStamped = TransformStamped()
        static_transformStamped.header.stamp = self.get_clock().now().to_msg() # 如何写入时间戳
        static_transformStamped.header.frame_id = 'world' # parent frame id
        static_transformStamped.child_frame_id = sys.argv[1] # child frame id
        # 填充六个位姿
        static_transformStamped.transform.translation.x = float(sys.argv[2])
        static_transformStamped.transform.translation.y = float(sys.argv[3])
        static_transformStamped.transform.translation.z = float(sys.argv[4])
        # 将欧拉表示成四元数
        quat = tf_transformations.quaternion_from_euler(
            float(sys.argv[5]), float(sys.argv[6]), float(sys.argv[7]))
        static_transformStamped.transform.rotation.x = quat[0]
        static_transformStamped.transform.rotation.y = quat[1]
        static_transformStamped.transform.rotation.z = quat[2]
        static_transformStamped.transform.rotation.w = quat[3]

        # 这个类里面有发布的对应方程
        self._tf_publisher.sendTransform(static_transformStamped)


def main():
    logger = rclpy.logging.get_logger('logger')

    # 从命令行获得参数
    if len(sys.argv) < 8:
        logger.info('Invalid number of parameters. Usage: \n'
                    '$ ros2 run turtle_tf2_py static_turtle_tf2_broadcaster'
                    'child_frame_name x y z roll pitch yaw')
        sys.exit(0)
    else:
        if sys.argv[1] == 'world':
            logger.info('Your static turtle name cannot be "world"')
            sys.exit(0)

    # pass parameters and initialize node
    rclpy.init()
    node = StaticFramePublisher(sys.argv)
    
    # 尝试spin 除非被叫停了
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    rclpy.shutdown()

File: ROS2/src/learning_tf2_cpp/src/static_turtle_tf2_broadcaster.cpp

/*
发布 static_transform_publsiher 有三种方法：
  一。 自定义
    可以自由实现功能，如四元数转换之类的
    ros2 run learning_tf2_cpp static_turtle_tf2_broadcaster mystaticturtle 0 0 1 0 0 0 （定义了parent frame 是world）
  二。使用自带的tf2_ros包
    旋转表示：
      ros2 run tf2_ros static_transform_publisher x y z yaw pitch roll frame_id child_frame_id
    四元数表达：
      ros2 run tf2_ros static_transform_publisher x y z qx qy qz qw frame_id child_frame_id
  三。ros2 launch
    ```
      from launch import LaunchDescription
      from launch_ros.actions import Node

      def generate_launch_description():
        return LaunchDescription([
            Node(
                  package='tf2_ros',
                  executable='static_transform_publisher',
                  arguments = ['0', '0', '1', '0', '0', '0', 'world', 'mystaticturtle']
            ),
        ])
    ```
*/






// 保函要发布的数据类型
#include <geometry_msgs/msg/transform_stamped.hpp>

//用到rclcpp里面的节点类
#include <rclcpp/rclcpp.hpp>

// 欧拉角和四元数之间的转换
#include <tf2/LinearMath/Quaternion.h>

#include <tf2_ros/static_transform_broadcaster.h>

#include <memory>

using std::placeholders::_1;

class StaticFramePublisher : public rclcpp::Node
{
public:
  explicit StaticFramePublisher(char * transformation[])
  : Node("static_turtle_tf2_broadcaster")
  {
    tf_publisher_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);

    // Publish static transforms once at startup
    this->make_transforms(transformation);
  }

private:
  void make_transforms(char * transformation[])
  {
    rclcpp::Time now = this->get_clock()->now();
    geometry_msgs::msg::TransformStamped t;

    t.header.stamp = now;
    t.header.frame_id = "world";
    t.child_frame_id = transformation[1];

    t.transform.translation.x = atof(transformation[2]);
    t.transform.translation.y = atof(transformation[3]);
    t.transform.translation.z = atof(transformation[4]);
    tf2::Quaternion q;
    q.setRPY(
      atof(transformation[5]),
      atof(transformation[6]),
      atof(transformation[7]));
    t.transform.rotation.x = q.x();
    t.transform.rotation.y = q.y();
    t.transform.rotation.z = q.z();
    t.transform.rotation.w = q.w();

    tf_publisher_->sendTransform(t);
  }
  std::shared_ptr<tf2_ros::StaticTransformBroadcaster> tf_publisher_;
};

int main(int argc, char * argv[])
{
  auto logger = rclcpp::get_logger("logger");

  // Obtain parameters from command line arguments
  if (argc != 8) {
    RCLCPP_INFO(
      logger, "Invalid number of parameters\nusage: "
      "ros2 run learning_tf2_cpp static_turtle_tf2_broadcaster "
      "child_frame_name x y z roll pitch yaw");
    return 1;
  }

  // As the parent frame of the transform is `world`, it is
  // necessary to check that the frame name passed is different
  if (strcmp(argv[1], "world") == 0) {
    RCLCPP_INFO(logger, "Your static turtle name cannot be 'world'");
    return 1;
  }

  // Pass parameters and initialize node
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<StaticFramePublisher>(argv));
  rclcpp::shutdown();
  return 0;
}
